src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -78,7 +86,29 @@ private static async Task WriteMessageCoreAsyncAwaited(this PipeWriter pipeWrite
 
         private static Task WriteMessageCoreAsync(this PipeWriter pipeWriter, byte[] messageData, HttpContextServerCallContext serverCallContext, bool flush)
         {
-            WriteHeader(pipeWriter, messageData.Length);
+            var canCompress = serverCallContext.CanWriteCompressed();
+
+            // If compression has not been disabled by WriteFlags.NoCompress then
+            //   1. Use compression algorithm of request
+            //   2. Use default compression algorithm configured on server
+            var compressionEncoding = canCompress
+                ? serverCallContext.GetGrpcEncoding()
```

Comment:
This doesn't address https://github.com/grpc/grpc-dotnet/pull/125#discussion_r268081897. Why is the encoding of the request used to compress the response?

Reply:
Actually I may be misunderstanding the flow chart a little.